src/main/java/com/hivemq/persistence/local/memory/ClientSessionMemoryLocalPersistence.java

```diff
@@ -0,0 +1,531 @@
+
+            if (ClientSessions.isExpired(clientSession, currentTimeMillis - timestamp)) {
+
+                if (sessionExpiryInterval > SESSION_EXPIRE_ON_DISCONNECT) {
+                    sessionsCount.decrementAndGet();
+                }
+                eventLog.clientSessionExpired(timestamp + sessionExpiryInterval * 1000, entry.getKey());
+                expiredClientIds.add(entry.getKey());
+                currentMemorySize.addAndGet(-storedEntry.getEstimatedSize());
+                iterator.remove();
+            }
+        }
+        return expiredClientIds.build();
+    }
+
+    @Override
+    public @NotNull Set<String> getDisconnectedClients(final int bucketIndex) {
+
+        final Map<String, PersistenceEntry<ClientSession>> bucket = getBucket(bucketIndex);
+        final long currentTimeMillis = System.currentTimeMillis();
+
+        return bucket.entrySet()
+                .stream()
+                .filter(entry -> !entry.getValue().getObject().isConnected())
+                .filter(entry -> entry.getValue().getObject().getSessionExpiryInterval() > 0)
+                .filter(entry -> {
+                    final PersistenceEntry<ClientSession> storedSession = entry.getValue();
+                    final long timeSinceDisconnect = currentTimeMillis - storedSession.getTimestamp();
+                    final long sessionExpiryIntervalInMillis =
+                            storedSession.getObject().getSessionExpiryInterval() * 1000L;
+                    return timeSinceDisconnect < sessionExpiryIntervalInMillis;
+                })
+                .map(Map.Entry::getKey)
+                .collect(ImmutableSet.toImmutableSet());
+    }
+
+    @Override
+    public int getSessionsCount() {
+        return sessionsCount.get();
+    }
+
+    @ExecuteInSingleWriter
+    @Override
+    public void setSessionExpiryInterval(
+            final @NotNull String clientId, final long sessionExpiryInterval, final int bucketIndex) {
+        checkNotNull(clientId, "Client Id must not be null");
+        ThreadPreConditions.startsWith(SINGLE_WRITER_THREAD_PREFIX);
+
+
+        if (sessionExpiryInterval < 0) {
+            throw new InvalidSessionExpiryIntervalException("Invalid session expiry interval " + sessionExpiryInterval);
+        }
+
+        final Map<String, PersistenceEntry<ClientSession>> bucket = getBucket(bucketIndex);
+
+        bucket.compute(clientId, (ignored, storedSession) -> {
+
+
+            if (storedSession == null) {
+                throw NoSessionException.INSTANCE;
+            }
+
+            final ClientSession clientSession = storedSession.getObject();
+
+            // is tombstone?
+            if (!clientSession.isConnected() && !persistent(clientSession)) {
+                throw NoSessionException.INSTANCE;
+            }
+
+            clientSession.setSessionExpiryInterval(sessionExpiryInterval);
+
+            return new PersistenceEntry<>(clientSession, System.currentTimeMillis());
```

Comment:
I thought the timestamp is only used as disconnected timestamp, why is it then set to the current timestamp here?

Reply:
its the connected and the disconnected timestamp, but yes setting it here is not correct.